### PR TITLE
Backend tests refactor

### PR DESCRIPTION
**Pull request by paolodamico**
## Changes

This PR refactors the way we handle test classes to optimize test running times, remove duplicated code and reduce confusion around which test classes to use.
- Chiefly, we reduce the multitude of classes and repeat setup code we had (distributed too in`posthog/api/test/base.py` & `posthog/test/base.py`) to leave only three test classes with specific use cases.
    - `BaseTest`. Based on Django's `TestCase`. Used for any unit tests or any general tests that do not test API endpoints. DB access is fully available.
    - `APIBaseTest`. Based on DRF's `APITestCase`. Used for any functional tests that test API endpoints.
    - `LicensedTestMixin`. Based on `APIBaseTest` (above). Introduces some extra setup for EE tests (such as setting up an enterprise `License` object. Should be used for any EE test.
- Tests now run based on Django's `TestCase` instead of `TransactionTestCase` to speed up test times (see below for details). `TestCase` wraps all tests and each test class in atomic blocks to avoid the need of recreating the test data on every test (more details on [the docs](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#testcase)).
- We introduce the `CLASS_DATA_LEVEL_SETUP` config (default = `True`). If set to `False` test data will be created on each test case instead of once per test class. This supports some particular test cases that do require test data to be setup on every test case.
- _Just as documentation_, there's some particular tests (e.g. `test_capture.py`, `test_decide.py`, ...) that while do they test the API, they a) test endpoints that do not run on DRF and b) test endpoints that require granular control over `Content-Type` sent. For these reasons, these test cases are run under the `BaseTest` class.
- Further logic simplification in test setup logic & cleanup.
- This PR does **not change any test or app logic except:**
    - Annotation functional test that wasn’t properly testing query string because it had an incorrect parameter name (tests were passing because we only had one test object).
    - Feature flag non-duplicate key validation now occurs at the serializer level, instead of relying on DB-level and catching `IntegrityException`s.

### Test speed-up results
_Results consider full test suite on Python 3.8.5 & Python 3.9.0_
- Overall GitHub Actions improvement (in run time) of over **30%**.
- PostHog tests now run over **50%** faster. Entire test suite now runs in about **65s** on GH Actions.
- EE tests now run over **25%** faster.

## Checklist

- [X] All querysets/queries filter by Organization, by Team, and by User. **Not applicable.**
- [X] Django backend tests
- [X] Jest frontend tests. **Not applicable.**
- [X] Cypress end-to-end tests. **Not applicable.**

### CODE REVIEW

**Comment by paolodamico** [comment]
PR is ready for review. Cloud tests are addressed in a separate PR on https://github.com/PostHog/posthog-cloud/

**Comment by paolodamico** [comment]
Cloud tests 100% ✅ (see https://github.com/PostHog/posthog-cloud/pull/99)!

File: posthog/api/feature_flag.py [hunk 1]
```diff
@@ -35,6 +38,15 @@ class Meta:
             "rollout_percentage",
         ]
 
+    def __init__(self, instance: Optional[Any] = None, data: Any = empty, **kwargs: Any):
```
Comment (macobo) [inline posthog/api/feature_flag.py:41]: This is suspect, I don't think we use this pattern anywhere else.

Is there a reason we need to override init over e.g. https://stackoverflow.com/questions/46173838/validate-specific-field-in-drf-serializer?

**Review by macobo — COMMENTED** [review]
Massive work and impact, though I wasn't aware it was an issue. Codewise this cleans a lot of things up which is a win in itself. One question only (below).

Long-term though we should start writing more tests using pytest (already introduced to the codebase).

File: posthog/api/feature_flag.py (hunk 1 above)
Comment (paolodamico) [reply posthog/api/feature_flag.py]: I've done this before in other projects and it works fine, but I opted for what you suggested.

**Pull request merged**